File: src/main/java/com/github/lppedd/cc/macro/CommitTypeMacro.java
package com.github.lppedd.cc.macro;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import com.github.lppedd.cc.CCBundle;
import com.github.lppedd.cc.lookup.CommitTypeLookupElement;
import com.github.lppedd.cc.psi.CommitTypePsiElement;
import com.intellij.codeInsight.lookup.LookupElement;
import com.intellij.codeInsight.template.Expression;
import com.intellij.codeInsight.template.ExpressionContext;
import com.intellij.codeInsight.template.Macro;
import com.intellij.codeInsight.template.Result;
import com.intellij.psi.PsiManager;

public class CommitTypeMacro extends Macro {
  private static final String NAME = "conventionalCommitType";

  @Override
  public String getName() {
    return NAME;
  }

  @Override
  public String getPresentableName() {
    return NAME + "()";
  }

  @Nullable
  @Override
  public Result calculateResult(
      @NotNull final Expression[] params,
      @Nullable final ExpressionContext context) {
    return null;
  }

  @Nullable
  @Override
  public LookupElement[] calculateLookupItems(
      @NotNull final Expression[] params,
      @Nullable final ExpressionContext context) {
    final PsiManager psiManager = PsiManager.getInstance(context.getProject());
    return new LookupElement[] {
        // @formatter:off
        new CommitTypeLookupElement(new CommitTypePsiElement(psiManager, "refactor", CCBundle.message("commit.type.refactor"))),
        new CommitTypeLookupElement(new CommitTypePsiElement(psiManager, "fix", CCBundle.message("commit.type.fix"))),
        new CommitTypeLookupElement(new CommitTypePsiElement(psiManager, "feat", CCBundle.message("commit.type.feat"))),
        new CommitTypeLookupElement(new CommitTypePsiElement(psiManager, "build", CCBundle.message("commit.type.build"))),
        new CommitTypeLookupElement(new CommitTypePsiElement(psiManager, "style", CCBundle.message("commit.type.style"))),
        new CommitTypeLookupElement(new CommitTypePsiElement(psiManager, "test", CCBundle.message("commit.type.test"))),
        new CommitTypeLookupElement(new CommitTypePsiElement(psiManager, "docs", CCBundle.message("commit.type.docs"))),
        new CommitTypeLookupElement(new CommitTypePsiElement(psiManager, "perf", CCBundle.message("commit.type.perf"))),
        new CommitTypeLookupElement(new CommitTypePsiElement(psiManager, "ci", CCBundle.message("commit.type.ci")))
        // @formatter:on
    };
  }
}

File: src/main/kotlin/com/github/lppedd/cc/macro/CommitTypeMacro.kt
package com.github.lppedd.cc.macro

import com.github.lppedd.cc.ConventionalCommitBundle
import com.github.lppedd.cc.lookup.CommitTypeLookupElement
import com.github.lppedd.cc.psi.CommitTypePsiElement
import com.intellij.codeInsight.lookup.LookupElement
import com.intellij.codeInsight.template.Expression
import com.intellij.codeInsight.template.ExpressionContext
import com.intellij.codeInsight.template.Macro
import com.intellij.psi.PsiManager

/**
 * @author Edoardo Luppi
 */
class CommitTypeMacro : Macro() {
  override fun getName() = NAME
  override fun getPresentableName() = "$NAME()"
  override fun calculateResult(
    params: Array<Expression>,
    context: ExpressionContext?
  ) = null

  override fun calculateLookupItems(
    params: Array<Expression>,
    context: ExpressionContext?
  ) = getTypes(context?.run { PsiManager.getInstance(project) })

  private fun getTypes(psiManager: PsiManager?): Array<LookupElement> {
    return if (psiManager == null) emptyArray()
    else arrayOf(
      CommitTypeLookupElement(1, CommitTypePsiElement("refactor", ConventionalCommitBundle["commit.type.refactor"], psiManager)),
      CommitTypeLookupElement(1, CommitTypePsiElement("fix", ConventionalCommitBundle["commit.type.fix"], psiManager)),
      CommitTypeLookupElement(1, CommitTypePsiElement("feat", ConventionalCommitBundle["commit.type.feat"], psiManager)),
      CommitTypeLookupElement(1, CommitTypePsiElement("build", ConventionalCommitBundle["commit.type.build"], psiManager)),
      CommitTypeLookupElement(1, CommitTypePsiElement("style", ConventionalCommitBundle["commit.type.style"], psiManager)),
      CommitTypeLookupElement(1, CommitTypePsiElement("test", ConventionalCommitBundle["commit.type.test"], psiManager)),
      CommitTypeLookupElement(1, CommitTypePsiElement("docs", ConventionalCommitBundle["commit.type.docs"], psiManager)),
      CommitTypeLookupElement(1, CommitTypePsiElement("perf", ConventionalCommitBundle["commit.type.perf"], psiManager)),
      CommitTypeLookupElement(1, CommitTypePsiElement("ci", ConventionalCommitBundle["commit.type.ci"], psiManager))
    )
  }

  companion object {
    private const val NAME = "conventionalCommitType"
  }
}
